apis/sgv2-quarkus-common/src/main/java/io/stargate/sgv2/api/common/optimistic/OptimisticQueryService.java

```diff
@@ -0,0 +1,179 @@
+   * @param queryFunction Function that creates the query from the {@link Schema.CqlTable}. Note
+   *     that the table can be <code>null</code>.
+   * @return Response when optimistic query is executed correctly.
+   */
+  public Uni<QueryOuterClass.Response> executeOptimistic(
+      String keyspace,
+      String table,
+      Function<String, Uni<? extends QueryOuterClass.Response>> missingKeyspace,
+      Function<Schema.CqlTable, Uni<QueryOuterClass.Query>> queryFunction) {
+
+    // get table from the schema manager without the hash validation
+    Uni<Schema.CqlKeyspaceDescribe> cqlUni = schemaManager.getKeyspace(keyspace, false);
+    return executeOptimistic(cqlUni, table, () -> missingKeyspace.apply(keyspace), queryFunction);
+  }
+
+  /**
+   * Executes the optimistic query against the keyspace provided in the given uni.
+   *
+   * @param cqlUni Uni providing the keyspace.
+   * @param table Table name.
+   * @param missingKeyspace Supplier in case the keyspace in case it's not existing. Usually there
+   *     to provide a failure.
+   * @param queryFunction Function that creates the query from the {@link Schema.CqlTable}. Note
+   *     that the table can be <code>null</code>.
+   * @return Response when optimistic query is executed correctly.
+   */
+  public Uni<QueryOuterClass.Response> executeOptimistic(
+      Uni<Schema.CqlKeyspaceDescribe> cqlUni,
+      String table,
+      Supplier<Uni<? extends QueryOuterClass.Response>> missingKeyspace,
+      Function<Schema.CqlTable, Uni<QueryOuterClass.Query>> queryFunction) {
+
+    // get from cql keyspace
+    return cqlUni
+
+        // if keyspace is found, execute optimistic query with that keyspace
+        .onItem()
+        .ifNotNull()
+        .transformToUni(
+            cqlKeyspace ->
+                executeWithKeyspace(cqlKeyspace, table, queryFunction)
+
+                    // once we have a result pipe to our handler
+                    .onItem()
+                    .transformToUni(handler(table, missingKeyspace, queryFunction)))
+
+        // if keyspace not found at first place, use mapper
+        .onItem()
+        .ifNull()
+        .switchTo(missingKeyspace.get());
+  }
+
+  // handles the QueryWithSchemaResponse
+  private Function<Schema.QueryWithSchemaResponse, Uni<? extends QueryOuterClass.Response>> handler(
+      String table,
+      Supplier<Uni<? extends QueryOuterClass.Response>> missingKeyspace,
+      Function<Schema.CqlTable, Uni<QueryOuterClass.Query>> queryFunction) {
+    return response -> {
+
+      // * if no keyspace, calls the missing key space function
+      // * if there is new keyspace, execute again with that keyspace, use same handler
+      // * if everything is fine, map to result
+      if (response.hasNoKeyspace()) {
+        return missingKeyspace.get();
+      } else if (response.hasNewKeyspace()) {
+        // note that this is endlessly trying to execute the query in case there is always a changes
```

Comment:
I think I'd be most worried about bug that indicates new keyspace returned. But to try to handle infinite repetition would probably require passing "number of attempts" via bridge to keep track... so not sure how to avoid.

Reply:
Yes, same thinking here, that's why I added the comment there.. Adding retry count is pritty simple, however, my main question would be what to do when we hit that last retry and we get a new keyspace then? Some kind of retries exhausted mapper than in addition?

Reply:
Right, good question. I think there should be an exception for failing to sync schema, plus count of retries that was used.